src/docker.cc

```diff
@@ -32,16 +34,91 @@ namespace google {
+constexpr const char kDockerContainerResourcePrefix[] = "container";
 
 }
 
 DockerReader::DockerReader(const MetadataAgentConfiguration& config)
     : config_(config), environment_(config) {}
 
+MetadataUpdater::ResourceMetadata DockerReader::GetContainerMetadata(
+    const json::Object* container, Timestamp collected_at) const
+    throw(json::Exception) {
+  const std::string zone = environment_.InstanceZone();
+
+  const std::string id = container->Get<json::String>("Id");
+  // Inspect the container.
+  try {
+    json::value raw_docker = QueryDocker(
+        std::string(kDockerEndpointPath) + "/" + id + "/json");
+    if (config_.VerboseLogging()) {
+      LOG(INFO) << "Parsed metadata: " << *raw_docker;
+    }
+
+    const json::Object* container_desc = raw_docker->As<json::Object>();
+    const std::string name = container_desc->Get<json::String>("Name");
+
+    const std::string created_str =
+        container_desc->Get<json::String>("Created");
+    Timestamp created_at = rfc3339::FromString(created_str);
+
+    const json::Object* state = container_desc->Get<json::Object>("State");
+    bool is_deleted = state->Get<json::Boolean>("Dead");
+
+    const MonitoredResource resource("docker_container", {
+      {"location", zone},
+      {"container_id", id},
+    });
+
+    json::value instance_resource =
+        InstanceReader::InstanceResource(environment_).ToJSON();
+
+    json::value raw_metadata = json::object({
+      {"blobs", json::object({
+        {"association", json::object({
+          {"version", json::string(config_.MetadataIngestionRawContentVersion())},
+          {"raw", json::object({
+            {"infrastructureResource", std::move(instance_resource)},
+          })},
+        })},
+        {"api", json::object({
+          {"version", json::string(config_.DockerApiVersion())},
+          {"raw", std::move(raw_docker)},
+        })},
+      })},
+    });
+    if (config_.VerboseLogging()) {
+      LOG(INFO) << "Raw docker metadata: " << *raw_metadata;
+    }
+
+    const std::string resource_id = boost::algorithm::join(
+        std::vector<std::string>{kDockerContainerResourcePrefix, id},
+        config_.MetadataApiResourceTypeSeparator());
+    const std::string resource_name = boost::algorithm::join(
+        // The container name reported by Docker will always have a leading '/'.
+        std::vector<std::string>{kDockerContainerResourcePrefix, name.substr(1)},
+        config_.MetadataApiResourceTypeSeparator());
+    return MetadataUpdater::ResourceMetadata(
+        std::vector<std::string>{resource_id, resource_name},
+        resource,
+#ifdef ENABLE_DOCKER_METADATA
+        MetadataAgent::Metadata(config_.MetadataIngestionRawContentVersion(),
+                                is_deleted, created_at, collected_at,
+                                std::move(raw_metadata))
+#else
+        MetadataAgent::Metadata::IGNORED()
+#endif
+    );
+  } catch (const QueryException& e) {
```

Comment:
Just a thought, now that we know what we know with the cpp-netlib, and that non 200 responses aren't actually throwing exceptions, the reason for this is more likely to be an issue with the docker daemon or the server it's running, not that the container disappears.